crowdsourcing/crowdsourcing_tutorial.py

```diff
@@ -260,39 +260,35 @@ def polarity_negative_2(x):
 # ## Use Soft Labels to Train End Model
 
 # %% [markdown]
-# For simplicity and speed, we use a simple "bag of n-grams" feature representation:
-# each data point is represented by a one-hot vector marking which words or 2-word
-# combinations are present in the comment text.
-
-# %% [markdown]
-# ### Featurization
+# ### Getting features from BERT
+# Since we have very limited training data, we cannot train a complex model like an LSTM with a lot of parameters. Instead, we use a pre-trained model, [BERT](https://github.com/google-research/bert), to generate embeddings for each our tweets, and treat the embedding values as features.
 
 # %%
-from sklearn.feature_extraction.text import CountVectorizer
+import torch
+from pytorch_transformers import BertModel, BertTokenizer
+import numpy as np
+
+model = BertModel.from_pretrained("bert-base-uncased")
+tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
 
-train_tokens = [row.tweet_text for _, row in df_train.iterrows()]
-test_tokens = [row.tweet_text for _, row in df_test.iterrows()]
 
-vectorizer = CountVectorizer(ngram_range=(1, 2))
-X_train = vectorizer.fit_transform(train_tokens).toarray().astype("float")
-X_test = vectorizer.transform(test_tokens).toarray().astype("float")
+def encode_text(text):
+    input_ids = torch.tensor([tokenizer.encode(text)])
+    return model(input_ids)[0].mean(1)[0].detach().numpy()
+
+
+train_vectors = np.array(list(df_train.tweet_text.apply(encode_text).values))
+test_vectors = np.array(list(df_test.tweet_text.apply(encode_text).values))
 
 # %% [markdown]
 # ### Model on soft labels
-# Now, we train a simple MLP model on the bag-of-words features, using labels
+# Now, we train a simple Logistic Regression model on the BERT features, using labels
 # obtained from our label model.
 
 # %%
-import tensorflow as tf
+from sklearn.linear_model import LogisticRegression
```

Comment:
could we do any better with a little regularization? just an idea, honestly idk

Reply:
Tried l1 and l2 regularization. l1 does worse, while l2 does the same.